pkg/microservice/aslan/core/environment/service/environment.go

```diff
@@ -1182,26 +1219,42 @@ func UpdateHelmProduct(productName, envName, updateType, username, requestID str
 		return e.ErrUpdateEnv.AddDesc(e.FindProductTmplErrMsg)
 	}
 
-	productRespMap := productResp.GetServiceMap()
+	deletedSvcSet := sets.NewString(deletedServices...)
+	// services need to be created or updated
+	serviceNeedUpdateOrCreate := sets.NewString()
+	for _, chart := range overrideCharts {
+		serviceNeedUpdateOrCreate.Insert(chart.ServiceName)
+	}
+	// use service definition from service template, but keep the image info
+	productServiceMap := productResp.GetServiceMap()
+	allServices := make([][]*commonmodels.ProductService, 0)
 	for _, svrs := range updateProd.Services {
+		svcGroup := make([]*commonmodels.ProductService, 0)
 		for _, svr := range svrs {
-			ps, ok := productRespMap[svr.ServiceName]
-			if !ok {
+			if deletedSvcSet.Has(svr.ServiceName) {
+				continue
+			}
+			ps, ok := productServiceMap[svr.ServiceName]
```

Comment:
Perhaps using a map to store the information in `ps` would improve the efficiency of the following lookup logic, for example, `map[serviceName][containerName]imageName`